### PR TITLE
Apply internal connection single node to analytics

### PR DESCRIPTION
**Pull request by qwu16**

### CODE REVIEW

File: source/agent/analytics/index.js [hunk 1]
```diff
@@ -95,65 +100,186 @@ module.exports = function (rpcClient, rpcId, agentId, clusterIp) {
...
+        }
+    };
+
+    that.cutoff = function (connectionId, callback) {
+        log.debug('cutoff, connectionId:', connectionId);
+        router.cutoff(connectionId).then(onSuccess(callback), onError(callback));
+    };
+
+    that.onFaultDetected = function (message) {
+        log.debug('onFaultDetected, message:', message);
```
Comment (taste1981) [inline source/agent/analytics/index.js:281]: invoke connections.onFaultDetected() ? and maybe exit current process?

File: source/agent/analytics/index.js [hunk 2]
```diff
@@ -95,65 +100,186 @@ module.exports = function (rpcClient, rpcId, agentId, clusterIp) {
...
+        });
+
+        return callback('callback', 'ok');
+      } else {
+        callback('callback', 'error', 'Connection already exist: '  + connectionId);
+      }
+    };
+
+    that.unsubscribe = function (connectionId, callback) {
+        log.debug('unsubscribe, connectionId:', connectionId);
```
Comment (taste1981) [inline source/agent/analytics/index.js:242]: Nothing to do here?

File: source/agent/analytics/index.js [hunk 3]
```diff
@@ -95,65 +100,186 @@ module.exports = function (rpcClient, rpcId, agentId, clusterIp) {
...
+      if(!inputs[connectionId]) {
+        const videoFormat = options.connection.video.format;
+        const videoParameters = options.connection.video.parameters;
+        const algo = options.connection.algorithm;
+
+        newStreamId = algo + options.media.video.from;
+
+        const pluginName = algorithms[algo].name;
+        let codec = videoFormat.codec;
+              if (videoFormat.profile) {
```
Comment (taste1981) [inline source/agent/analytics/index.js:162]: spacing.

File: source/agent/analytics/index.js [hunk 4]
```diff
@@ -95,65 +100,186 @@ module.exports = function (rpcClient, rpcId, agentId, clusterIp) {
...
+                codec += '_' + videoFormat.profile;
+              }
+        codec = codec.toLowerCase();
+        const {resolution, framerate, keyFrameInterval, bitrate}
+                = getVideoParameterForAddon(options.connection.video);
```
Comment (starwarfan) [inline source/agent/analytics/index.js:167]: space

File: source/agent/analytics/index.js [hunk 5]
```diff
@@ -95,65 +100,186 @@ module.exports = function (rpcClient, rpcId, agentId, clusterIp) {
...
+        }
+        inputs[connectionId] = true;
+
+        engine.addEventListener('fatal', function (error) {
+            log.error('GStreamer pipeline error:', error);
+            notifyStatus(options.controller, connectionId, 'out', {type: 'failed', reason: 'Analytics error: ' + error});
+        });
+
+        const streamInfo = {
+          type: 'analytics',
```
Comment (starwarfan) [inline source/agent/analytics/index.js:190]: use same tab size, this is 2 spaces, but some other places are 4 spaces.

File: source/agent/analytics/index.js [hunk 6]
```diff
@@ -95,65 +100,186 @@ module.exports = function (rpcClient, rpcId, agentId, clusterIp) {
...
+
+    that.unsubscribe = function (connectionId, callback) {
+        log.debug('unsubscribe, connectionId:', connectionId);
+    };
+
+    // streamInfo = {id: 'string', ip: 'string', port: 'number'}
+    // from = {audio: streamInfo, video: streamInfo, data: streamInfo}
+    that.linkup = function (connectionId, fromInfo, callback) {
+        log.debug('linkup, connectionId:', connectionId, 'from:', fromInfo);
+        if (inputs[connectionId]) {
```
Comment (starwarfan) [inline source/agent/analytics/index.js:249]: inputs[connectionId] share  the same engine?
Only 1 engine for multiple inputs? or just 1 input for 1 process?

Comment (qwu16) [reply source/agent/analytics/index.js]: only 1 input for 1 process

File: source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc [hunk 7]
```diff
@@ -273,17 +283,23 @@ int VideoGstAnalyzer::createPipeline()
...
 
     loop = g_main_loop_new(NULL, FALSE);
 
     m_bus = gst_pipeline_get_bus(GST_PIPELINE(pipeline));
     m_bus_watch_id = gst_bus_add_watch(m_bus, StreamEventCallBack, this);
+
+    rvaStatus status = pipeline_->LinkElements();
+    if(status != RVA_ERR_OK) {
+       ELOG_ERROR("Link element failed with rvastatus:%d\n",status);
+       return -1;
```
Comment (taste1981) [inline source/agent/analytics/videoGstPipeline/VideoGstAnalyzer.cc:297]: return false.

**Review by starwarfan — APPROVED** [review]

**Review by taste1981 — APPROVED** [review]

**Pull request merged**